### PR TITLE
Blog - Deploying Docker Containers to AWS ECS Using Terraform

### PR DESCRIPTION
**Pull request by msali123**
Done with all checks.

### CODE REVIEW

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+COPY . .
+~~~
+
+Add a port number to expose the Docker image:
+
+~~~{.bash caption=">_"}
+# Port number to expose the Node.js app outside of Docker
+EXPOSE 5000
+
+~~~
```
Comment (jalletto) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:163]: Captions for all of these should be "Dockerfile"

**Review by jalletto — CHANGES_REQUESTED** [review]

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md (hunk 1 above)
Comment (msali123) [reply blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md]: @jalletto done

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+Terraform checks the application's Dockerfile and instructs Docker on what needs to be done. Dockerfile contains the instructions used to create a Docker image. To create a docker image for the application, follow the steps outlined below:
+
+First, create a `Dockerfile` file in the `demo_node_app` directory. Inside this file, add the following:
+
+A [DockerHub](https://hub.docker.com/_/node) image to run Node.js:
+
+~~~{.bash caption=">_"}
+# Pull the Node.js image
+FROM node:18-alpine
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:126]:
```suggestion

~~~{.dockerfile caption="Dockerfile"}
# Pull the Node.js image
FROM node:18-alpine
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+
+Create a directory to host the application on the Docker:
+
+~~~{.bash caption="Dockerfile"}
+# Create a Docker working directory
+WORKDIR /app
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:133]:
```suggestion
~~~{.dockerfile caption="Dockerfile"}
# Create a Docker working directory
WORKDIR /app
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+
+Copy all the dependency files:
+
+~~~{.bash caption="Dockerfile"}
+# Copy package.json and package-lock.json dependencies files 
+COPY package*.json ./
+
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:141]:
```suggestion
~~~{.dockerfile caption="Dockerfile"}
# Copy package.json and package-lock.json dependencies files 
COPY package*.json ./

~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+
+Install the application's dependencies:
+
+~~~{.bash caption="Dockerfile"}
+# Install dependencies inside Docker
+RUN npm install
+~~~
+
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:149]:
```suggestion

~~~{.dockerfile caption="Dockerfile"}
# Install dependencies inside Docker
RUN npm install
~~~

```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 6]
```diff
@@ -0,0 +1,686 @@
...
+Copy the application files to the Docker directory:
+
+~~~{.bash caption="Dockerfile"}
+# Copy the application source code
+COPY . .
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:155]:
```suggestion

~~~{.dockerfile caption="Dockerfile"}
# Copy the application source code
COPY . .
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 7]
```diff
@@ -0,0 +1,686 @@
...
+
+Add a port number to expose the Docker image:
+
+~~~{.bash caption="Dockerfile"}
+# Port number to expose the Node.js app outside of Docker
+EXPOSE 5000
+
+~~~
+
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:164]:
```suggestion

~~~{.dockerfile caption="Dockerfile"}
# Port number to expose the Node.js app outside of Docker
EXPOSE 5000

~~~

```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 8]
```diff
@@ -0,0 +1,686 @@
...
+Finally, add the command to run the application:
+
+~~~{.bash caption=">_"}
+# Command to run the application
+CMD ["node", "index.js"]
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:170]:
```suggestion

~~~{.dockerfile caption="Dockerfile"}
# Command to run the application
CMD ["node", "index.js"]
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 9]
```diff
@@ -0,0 +1,686 @@
...
+terraform {
+  required_providers {
+    aws = {
+      source = "hashicorp/aws"
+      version = "4.45.0"
+    }
+  }
+}
+~~~
+
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:211]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
      version = "4.45.0"
    }
  }
}
~~~

```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+~~~{.bash caption="main.tf"}
+# main.tf
+provider "aws" {
+  region  = "us-east-1" #The region where the environment 
+  #is going to be deployed # Use your own region here
+  access_key = "enter_access_key_here" # Enter AWS IAM 
+  secret_key = "enter_secret_key_here" # Enter AWS IAM 
+}
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:222]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
provider "aws" {
  region  = "us-east-1" #The region where the environment 
  #is going to be deployed # Use your own region here
  access_key = "enter_access_key_here" # Enter AWS IAM 
  secret_key = "enter_secret_key_here" # Enter AWS IAM 
}
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+
+Finally, create an ECR using Terraform as follows:
+
+~~~{.bash caption="main.tf"}
+# main.tf
+resource "aws_ecr_repository" "app_ecr_repo" {
+  name = "app-repo"
+}
+~~~
+
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:232]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
resource "aws_ecr_repository" "app_ecr_repo" {
  name = "app-repo"
}
~~~

```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 12]
```diff
@@ -0,0 +1,686 @@
...
+So far, you've created a repository and deployed the image. But whenever you want to launch, you'll need a **target**. A cluster acts as the container target. It takes a task into the cluster configuration and runs that task within the cluster.
+The ECS agent communicates with the ECS cluster and receives requests to launch the container. To create a cluster where you'll run your task, add the following configurations to your `main.tf` file:
+
+~~~{caption="main.tf"}
+# main.tf
+resource "aws_ecs_cluster" "my_cluster" {
+  name = "app-cluster" # Name your cluster here
+}
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:330]:
```suggestion
~~~{.terraform caption="main.tf"}
# main.tf
resource "aws_ecs_cluster" "my_cluster" {
  name = "app-cluster" # Name your cluster here
}
~~~
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 13]
```diff
@@ -0,0 +1,686 @@
...
+
+An AWS ECS workload has two main launch types: [EC2](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-anywhere.html) and [Fargate](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/AWS_Fargate.html).
+
+Fargate is an AWS orchestration tool. It allows you to give AWS the role of managing your container lifecycle and the hosting infrastructure. Fargate runs your container as serverless, which means you don't need to provision your container using EC2 instances. Instead of using the ECS clusters to run your container, you can use Fargate to spin up and run your container on ECS without provisioning a virtual machine on AWS.
+
+In the next steps, you'll write the configuration that a task requires to spin up your Docker container. Terraform will automate all this for you.
+
+You need to build a task definition to get the application ready to operate on ECS. The task specification is a text file in JSON format that lists one or more resources that make up your container. Using Task definition JSON format, provide the container specifications as follows:
+
+~~~{ caption="main.tf"}
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:363]:
```suggestion

~~~{.terraform caption="main.tf"}
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 14]
```diff
@@ -0,0 +1,686 @@
...
+}
+~~~
+
+As described in the above config block, Terraform will create a task named `app-first-task` and also assign the resources needed to power up the container through this task. This process includes assigning the deployed image, container ports, launch type, and the hardware requirements that the container needs to run.
+
+Creating a task definition requires `ecsTaskExecutionRole` to be added to your IAM. The above task definition needs this role, and it is specified as `aws_iam_role.ecsTaskExecutionRole.arn`. In the next step, create a resource to execute this role as follows:
+
+~~~{ caption="main.tf"}
+# main.tf
+resource "aws_iam_role" "ecsTaskExecutionRole" {
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:398]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
resource "aws_iam_role" "ecsTaskExecutionRole" {
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 15]
```diff
@@ -0,0 +1,686 @@
...
+First, you need to create a [Virtual Private Cloud Module (VPC)](https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/latest) and subnet to launch your cluster into. VPC and subnet allow you to connect to the internet, communicate with ECS, and expose the application to available zones.
+
+Go ahead and create a default [VPC](https://earthly.dev/blog/aws-networks/#vpcs) and [subnets](https://earthly.dev/blog/aws-networks/#subnets) information for your AWS availability zones.
+
+~~~{ caption="main.tf"}
+# main.tf
+# Provide a reference to your default VPC
+resource "aws_default_vpc" "default_vpc" {
+}
+
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:471]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
# Provide a reference to your default VPC
resource "aws_default_vpc" "default_vpc" {
}

```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 16]
```diff
@@ -0,0 +1,686 @@
...
+}
+~~~
+
+### Implement a Load Balancer
+
+Next, create a security group that will route the HTTP traffic using a load balancer. Go ahead and implement a load balancer as follows:
+
+~~~{ caption="main.tf"}
+# main.tf
+resource "aws_alb" "application_load_balancer" {
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:490]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
resource "aws_alb" "application_load_balancer" {
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 17]
```diff
@@ -0,0 +1,686 @@
...
+~~~
+
+The above configuration creates a load balancer that will distribute the workloads across multiple resources to ensure application's availability, scalability, and security.
+
+### Creating a Security Group for the Load Balancer
+
+The next important part of allowing HTTP traffic to access the [ECS](/blog/how-to-setup-and-use-amazons-elastic-container-registry) cluster is to create a **security group**. This will be crucial for accessing the application later in this guide. Go ahead and add the security group for the load balancer as follows:
+
+~~~{ caption="main.tf"}
+# main.tf
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:509]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 18]
```diff
@@ -0,0 +1,686 @@
...
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+~~~
+
+Configure the load balancer with the VPC networking we created earlier. This will distribute the balancer traffic to the available zone:
+
+~~~{ caption="main.tf"}
+# main.tf
+resource "aws_lb_target_group" "target_group" {
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:532]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
resource "aws_lb_target_group" "target_group" {
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 19]
```diff
@@ -0,0 +1,686 @@
...
+  }
+}
+~~~
+
+### Create an ECS Service
+
+The last step is to create an ECS Service and its details to maintain task definition in an Amazon ECS cluster. The service will run the [cluster, task, and Fargate](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/AWS_Fargate.html) behind the created load balancer to distribute traffic across the containers that are associated with the service. You can achieve this with the following block:
+
+~~~{ caption="main.tf"}
+# main.tf
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:556]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 20]
```diff
@@ -0,0 +1,686 @@
...
+    assign_public_ip = true     # Provide the containers with public IPs
+    security_groups  = ["${aws_security_group.service_security_group.id}"] # Set up the security group
+  }
+}
+~~~
+
+To access the ECS service over HTTP while ensuring the VPC is more secure, create security groups that will only allow the traffic from the created load balancer. To do so, create a `aws_security_group.service_security_group` resource as follows:
+
+~~~{ caption="main.tf"}
+# main.tf
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:581]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
```

File: blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md [hunk 21]
```diff
@@ -0,0 +1,686 @@
...
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+~~~
+
+Additionally, add an output config that will extract the load balancer URL value from the state file and log it onto the terminal.
+
+~~~{ caption="main.tf"}
+# main.tf
```
Comment (balapriyac) [inline blog/_posts/2023-01-31-deploy-dockcontainers-to-AWSECS-using-Terraform.md:603]:
```suggestion

~~~{.terraform caption="main.tf"}
# main.tf
```

**Review by balapriyac — CHANGES_REQUESTED** [review]
@jalletto I've finished reviewing this PR.

**Comment by msali123** [comment]
@balapriyac Requested changes done.

**Pull request merged**